File: api/asset/dirscan.py
# -------------------------------------
# @file      : dirscan.py
# @author    : Autumn
# @contact   : rainy-autumn@outlook.com
# @time      : 2024/5/9 20:22
# -------------------------------------------

from bson import ObjectId
from fastapi import APIRouter, Depends
from motor.motor_asyncio import AsyncIOMotorCursor
from api.users import verify_token
from core.db import get_mongo_db
from core.util import search_to_mongodb, get_search_query
from loguru import logger

router = APIRouter()


@router.post("/data")
async def dirscan_data(request_data: dict, db=Depends(get_mongo_db), _: dict = Depends(verify_token)):
    try:
        page_index = request_data.get("pageIndex", 1)
        page_size = request_data.get("pageSize", 10)
        query = await get_search_query("DirScanResult", request_data)
        if query == "":
            return {"message": "Search condition parsing error", "code": 500}
        total_count = await db['DirScanResult'].count_documents(query)
        sort = request_data.get("sort", {})
        sort_by = [('_id', -1)]
        if sort != {}:
            if 'length' in sort:
                sort_value = sort['length']
                if sort_value is not None:
                    if sort_value == "ascending":
                        sort_value = 1
                    else:
                        sort_value = -1
                    sort_by = [('length', sort_value)]
        cursor: AsyncIOMotorCursor = ((db['DirScanResult'].find(query, {"_id": 0, "id": {"$toString": "$_id"}, "url": 1,
                                                                        "status": 1, "msg": 1, "length": 1, "tags": 1})
                                       .sort(sort_by)
                                       .skip((page_index - 1) * page_size)
                                       .limit(page_size)))
        result = await cursor.to_list(length=None)
        return {
            "code": 200,
            "data": {
                'list': result,
                'total': total_count
            }
        }
    except Exception as e:
        logger.error(str(e))
        # Handle exceptions as needed
        return {"message": "error", "code": 500}

File: internal/services/assets/dirscan/dirscan.go
// dirscan-------------------------------------
// @file      : dirscan.go
// @author    : Autumn
// @contact   : rainy-autumn@outlook.com
// @time      : 2025/5/17 15:12
// -------------------------------------------

package dirscan

import (
	"context"
	"github.com/Autumn-27/ScopeSentry-go/internal/models"
	"github.com/Autumn-27/ScopeSentry-go/internal/repositories/assets/dirscan"
	"github.com/Autumn-27/ScopeSentry-go/internal/utils/helper"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type Service interface {
	List(ctx context.Context, query models.SearchRequest) ([]models.DirScanResult, error)
}

type service struct {
	repo dirscan.Repository
}

func NewService() Service {
	return &service{repo: dirscan.NewRepository()}
}

func (s *service) List(ctx context.Context, query models.SearchRequest) ([]models.DirScanResult, error) {
	query.Index = "DirScanResult"
	searchQuery, err := helper.GetSearchQuery(query)
	if err != nil {
		return nil, err
	}
	filter := bson.M(searchQuery)
	sortBy := bson.D{{Key: "_id", Value: -1}}
	if val, ok := query.Sort["length"]; ok {
		dir := -1
		if val == "ascending" {
			dir = 1
		}
		sortBy = bson.D{{Key: "length", Value: dir}}
	}

	opts := options.Find().
		SetProjection(bson.M{
			"_id":    1,
			"url":    1,
			"status": 1,
			"msg":    1,
			"length": 1,
			"tags":   1,
		}).
		SetSort(sortBy).
		SetSkip(int64((query.PageIndex - 1) * query.PageSize)).
		SetLimit(int64(query.PageSize))

	return s.repo.FindWithPagination(ctx, filter, opts)
}
